### ISSUE TITLE
C++ member function ABI on Windows is not compatible with COM member functions.

### ISSUE THREAD

**Issue by moonheart08** [shader-slang/slang#5566]
This manifests itself if you attempt to call COM functions or provide your own. Every argument after `this` is mangled due to mismatching ABI. COM-aware bindings generators (like ClangSharpPInvokeGenerator) will use Slang's choice of calling convention, but at this time there is a mismatch between Slang's IUnknown and the actual IUnknown that is masked by the fact neither have any member functions with arguments.

On Windows, the C++ instance member function calling convention for cdecl and stdcall is not the same as the non-member function calling convention.

**Comment by csyonghe** [comment]
It seems like we need to remove all the SLANG_MCALL declarations on the COM interface types. Is this something you can help us clean up?

Btw, if you are working on creating a C# binding, would you mind upstream your work to this repo?

**Comment by moonheart08** [comment]
> It seems like we need to remove all the SLANG_MCALL declarations on the COM interface types. Is this something you can help us clean up?
> 
> Btw, if you are working on creating a C# binding, would you mind upstream your work to this repo?

I don't have slang's build environment set up locally at this time and I'm focused on another project (the one that lead to me finding these issues). Bindings were being produced as part of that project.

I will definitely consider upstreaming the bindings at some future point given I work out all the issues.

**Comment by csyonghe** [comment]
How can we help get you unblocked? It shouldn't take too much effort to make all the interfaces COM compatible, is this enough? The C binding part is more involved and I am not sure we have capacity to get to that soon and that would be something we could really appreciate community help.

**Comment by moonheart08** [comment]
> How can we help get you unblocked? It shouldn't take too much effort to make all the interfaces COM compatible, is this enough? The C binding part is more involved and I am not sure we have capacity to get to that soon and that would be something we could really appreciate community help.

At this time I'm just not interested in working further on it. I will likely give it another shot at some future (undetermined) date, but I'm focused on pushing my current project forward as my main priority right now. As for which issues I consider blockers on any future work, they are this issue and #5562. The C header issue is not pressing due to the binding generator(s) being used being capable of understanding C++, but may impact other integrators.

**Comment by csyonghe** [comment]
Thank you for sharing this feedback. We will prioritize fixing this and 5562 accordingly. Let us know if you decide to give the c# binding another try in the future!

**Comment by expipiplus1** [comment]
@moonheart08 what calling convention specifically are you expecting to see for these functions? I was under the impression that on x86_64 we are already doing the correct thing. Would changing it to whatever `STDMETHODCALLTYPE` expands to be sufficient for you?

**Comment by moonheart08** [comment]
> @moonheart08 what calling convention specifically are you expecting to see for these functions? I was under the impression that on x86_64 we are already doing the correct thing. Would changing it to whatever `STDMETHODCALLTYPE` expands to be sufficient for you?

This sounds correct.

**Comment by csyonghe** [comment]
@moonheart08 Currently all the com methods are marked as SLANG_MCALL, which IIRC is equivalent to STDMETHODCALLTYPE: stdcall on x86, and cdecl on x64. Are we missing anything here?

**Comment by tannergooding** [comment]
COM instance methods are notably not just `STDCALL` but rather a hybrid of `THISCALL+STDCALL`

For the most part these look and operate identically, but actual instance functions on Windows have a slightly different behavior when it comes to struct returns. In particular, `int` and `struct Wrapper { int value; }` are considered distinct for `THISCALL+STDCALL` where-as they are treated the same for just `STDCALL`.

Historically this has surfaced to users for APIs where the C++ definition is returning a struct that contains a singular primitive field, such as for `D3D12_GPU_DESCRIPTOR_HANDLE`. This resulted in developers needing to manually fix some COM APIs to handle the hidden return buffer parameter that is part of the underlying ABI. However, while this generally worked for x86/x64, it was incorrect in some cases for Arm64 or other platforms.

.NET added `CallConvMemberFunction` as a way to properly handle the ABI for COM like functions. However, since it handles this properly and struct wrappers are not equivalent it means that if you've defined your own "helper wrappers" on the .NET side, there will be a signature mismatch as well. This most commonly comes about due to defining something like `struct HRESULT { int value; }` where-as the C++ definition is strictly `typedef int HRESULT;` and thus is actually a primitive int.

ClangSharp will handle it downlevel by assuming you're only targeting x86/x64 and will do the relevant fixups using `STDCALL`. Likewise on newer versions of .NET where `CallConvMemberFunction` exists it will do the right thing here as well, including adding helper functions to allow wrapper types to be exposed for managed users while still allowing the underlying vtbl entries to be correct and take the raw primitives. An example of this would be for `ID3D12DescriptorHeap` where it has functions handling both types of cases: https://source.terrafx.dev/#TerraFX.Interop.Windows/DirectX/d3d12/d3d12/ID3D12DescriptorHeap.cs,e93a6a41a828d77a,references

**Comment by moonheart08** [comment]
Seems to have managed to be a me issue, thankfully!

I apologize for the wasted time here, ABIs are hard.

Underlying issue: In a few places, in place of returning int32_t for slangresult, I instead inserted my wrapper struct, which is not equivalent ABI wise as noted above! Good insights all around, I'm going to pull my slang wrappers back up from my work branch and see what I can do with them given this has been figured out.
